presidio-image-redactor/presidio_image_redactor/image_processing_engine.py

```diff
@@ -0,0 +1,361 @@
+        self,
+        preprocessing_kwargs: Optional[dict] = None,
+    ) -> None:
+        """Initialize the class.
+
+        :param preprocessing_kwargs: Optional preprocessing keyword arguments.
+        """
+
+        self.preprocessing_kwargs = preprocessing_kwargs
+
+    def preprocess_image(
+        self, image: PngImagePlugin.PngImageFile
+    ) -> Tuple[Image.Image, dict]:
+        """Preprocess the image to be analyzed.
+
+        :param image: Loaded PIL image.
+
+        :return: The processed image and any metadata regarding the
+             preprocessing approach.
+        """
+        return image, {}
+
+    def convert_image_to_array(self, image: PngImagePlugin.PngImageFile) -> np.ndarray:
+        """Convert PIL image to numpy array.
+
+        :param image: Loaded PNG image.
+
+        :return: image pixels as a numpy array.
+
+        """
+
+        if isinstance(image, np.ndarray):
+            img = image
+        else:
+            image = image.convert("L")
+            img = np.asarray(image)
+        return img
+
+    @staticmethod
+    def _get_bg_color(
+        image: PngImagePlugin.PngImageFile, is_greyscale: bool, invert: bool = False
+    ) -> Union[int, Tuple[int, int, int]]:
+        """Select most common color as background color.
+
+        :param image: Loaded PNG image.
+        :param is_greyscale: Whether the image is greyscale.
+        :param invert: TRUE if you want to get the inverse of the bg color.
+
+        :return: Background color.
+        """
+        # Invert colors if invert flag is True
+        if invert:
+            if image.mode == "RGBA":
+                # Handle transparency as needed
+                r, g, b, a = image.split()
+                rgb_image = Image.merge("RGB", (r, g, b))
+                inverted_image = PIL.ImageOps.invert(rgb_image)
+                r2, g2, b2 = inverted_image.split()
+
+                image = Image.merge("RGBA", (r2, g2, b2, a))
+
+            else:
+                image = PIL.ImageOps.invert(image)
+
+        # Get background color
+        if is_greyscale:
+            # Select most common color as color
+            bg_color = int(np.bincount(image.flatten()).argmax())
+        else:
+            # Reduce size of image to 1 pixel to get dominant color
+            tmp_image = image.copy()
+            tmp_image = tmp_image.resize((1, 1), resample=0)
+            bg_color = tmp_image.getpixel((0, 0))
+
+        return bg_color
+
+    @staticmethod
+    def _get_image_contrast(image: np.ndarray) -> Tuple[str, float]:
+        """Compute the contrast level and mean intensity of an image.
+
+        :param image: Input image pixels (as a numpy array).
+
+        :return: A tuple containing the contrast level and mean intensity of the image.
+        """
+        contrast = np.std(image)
+        mean_intensity = np.mean(image)
+        return contrast, mean_intensity
+
+
+class BilateralFilter(ImagePreprocessor):
+    """BilateralFilter class."""
```

Comment:
Would you mind adding more detail in the docstring for the class, just to provide an overview of what this class is/does?

Reply:
**nit**: Typo/formatting issue with the newly added description.